pallets/pallet-asset-swap/src/lib.rs

```diff
@@ -0,0 +1,548 @@
+			);
+
+			// 3. Verify the tx submitter has enough local assets for the swap, without
+			//    having their balance go to zero.
+			T::LocalCurrency::can_withdraw(&submitter, local_asset_amount)
+				.into_result(true)
+				.map_err(|_| DispatchError::from(Error::<T>::UserSwapBalance))?;
+
+			// 4. Verify the local assets can be transferred to the swap pool account
+			T::LocalCurrency::can_deposit(&swap_pair.pool_account, local_asset_amount, Provenance::Extant)
+				.into_result()
+				.map_err(|_| DispatchError::from(Error::<T>::LocalPoolBalance))?;
+
+			// 5. Verify we have enough balance on the remote location to perform the
+			//    transfer
+			let remote_asset_amount_as_u128 = local_asset_amount.into();
+			ensure!(
+				swap_pair.remote_asset_balance >= remote_asset_amount_as_u128,
+				Error::<T>::Liquidity
+			);
+
+			let asset_id_v3: AssetId = swap_pair.remote_asset_id.clone().try_into().map_err(|e| {
+				log::error!(
+					target: LOG_TARGET,
+					"Failed to convert asset ID {:?} into v3 `AssetId` with error {:?}",
+					swap_pair.remote_asset_id,
+					e
+				);
+				DispatchError::from(Error::<T>::Internal)
+			})?;
+			let remote_asset_fee_v3: MultiAsset = swap_pair.remote_fee.clone().try_into().map_err(|e| {
+				log::error!(
+					target: LOG_TARGET,
+					"Failed to convert remote XCM asset fee {:?} into v3 `MultiAssset` with error {:?}",
+					swap_pair.remote_fee,
+					e
+				);
+				DispatchError::from(Error::<T>::Xcm)
+			})?;
+			let destination_v3: MultiLocation = swap_pair.remote_reserve_location.clone().try_into().map_err(|e| {
+				log::error!(
+					target: LOG_TARGET,
+					"Failed to convert remote reserve location {:?} into v3 `MultiLocation` with error {:?}",
+					swap_pair.remote_reserve_location,
+					e
+				);
+				DispatchError::from(Error::<T>::Internal)
+			})?;
+			let beneficiary_v3: MultiLocation = (*beneficiary.clone()).try_into().map_err(|e| {
+				log::info!(
+					target: LOG_TARGET,
+					"Failed to convert beneficiary {:?} into v3 `MultiLocation` with error {:?}",
+					beneficiary,
+					e
+				);
+				DispatchError::from(Error::<T>::Xcm)
+			})?;
+			// Use the same local `AccountIdConverter` to generate a `MultiLocation` to use
+			// to send funds on remote.
+			let submitter_as_multilocation = T::AccountIdConverter::try_convert(submitter.clone())
+				.map(|j| j.into_location())
+				.map_err(|e| {
+					log::info!(
+						target: LOG_TARGET,
+						"Failed to convert account {:?} into `MultiLocation` with error {:?}",
+						submitter,
+						e
+					);
+					DispatchError::from(Error::<T>::Xcm)
+				})?;
+
+			// 6. Compose and validate XCM message
+			let appendix: Xcm<()> = vec![
+				RefundSurplus,
+				DepositAsset {
+					assets: MultiAssetFilter::Wild(WildMultiAsset::All),
+					beneficiary: submitter_as_multilocation,
+				},
+			]
+			.into();
+			let remote_xcm: Xcm<()> = vec![
+				WithdrawAsset(remote_asset_fee_v3.clone().into()),
```

Comment:
I guess on AH we are only allowed to pay the fees with dots. I think we have to still withdraw those assets.

Reply:
No, you can also pay with other assets. But here the `remote_asset_fee_v3` is set by KILT governance when setting a swap pair, and in the case of assets bridged to AH, that will be indeed the DOT token. But if the token is wrapped on a different chain, the asset that will be used for the fees might be different than the DOT. The pallet is not coupled to a specific asset.